lightningd/peer_htlcs.c

```diff
@@ -91,26 +95,35 @@ static u8 *make_failmsg(const tal_t *ctx, const struct htlc_in *hin,
 		return towire_final_incorrect_cltv_expiry(ctx, 0);
 	case WIRE_FINAL_INCORRECT_HTLC_AMOUNT:
 		return towire_final_incorrect_htlc_amount(ctx, hin->msatoshi);
+	case WIRE_INVALID_ONION_VERSION:
+	case WIRE_INVALID_ONION_HMAC:
+	case WIRE_INVALID_ONION_KEY:
+		log_broken(hin->key.peer->log, "Bad failmsg for %s",
+			   onion_type_name(failcode));
 	}
 	abort();
 }
 
-static void fail_htlc(struct htlc_in *hin, enum onion_type failcode,
-		      const struct sha256 *onion_sha)
+static void fail_htlc(struct htlc_in *hin, enum onion_type failcode)
 {
-	u8 *msg;
-
-	log_broken(hin->key.peer->log, "failed htlc %"PRIu64" code 0x%04x (%s)",
-		   hin->key.id, failcode, onion_type_name(failcode));
+	log_info(hin->key.peer->log, "failed htlc %"PRIu64" code 0x%04x (%s)",
+		 hin->key.id, failcode, onion_type_name(failcode));
 
-	if (failcode & UPDATE) {
-		/* FIXME: Ask gossip daemon for channel_update. */
-	}
+	if (failcode & BADONION)
+		hin->malformed = failcode;
+	else {
+		u8 *msg;
 
-	msg = make_failmsg(hin, hin, failcode, onion_sha, NULL);
-	hin->failuremsg = create_onionreply(hin, &hin->shared_secret, msg);
-	tal_free(msg);
+		if (failcode & UPDATE) {
+			/* FIXME: Ask gossip daemon for channel_update. */
```

Comment:
Yes, that's separate from this.  We have a FIXME for that too:

	/* FIXME: check for routing failure / perm fail. */
	/* check_for_routing_failure(i, sender, failure_code); */